x/ibc/02-client/types/tendermint/misbehaviour.go

```diff
@@ -1,96 +1,86 @@
 }
 
-// Type implements exported.Evidence interface
-func (ev Evidence) Route() string {
-	return exported.ClientTypeTendermint
+// ClientType is Tendermint light client
+func (m Misbehaviour) ClientType() exported.ClientType {
+	return exported.Tendermint
 }
 
-// Type implements exported.Evidence interface
-func (ev Evidence) Type() string {
-	return exported.ClientTypeTendermint
+// GetEvidence returns the evidence to handle a light client misbehaviour
+func (m Misbehaviour) GetEvidence() evidenceexported.Evidence {
+	return m.Evidence
 }
 
-// String implements exported.Evidence interface
-func (ev Evidence) String() string {
-	bz, err := yaml.Marshal(ev)
-	if err != nil {
-		panic(err)
+// ValidateBasic performs the basic validity checks for the evidence and the
+// client ID.
+func (m Misbehaviour) ValidateBasic() error {
+	if m.Evidence == nil {
+		return errors.ErrInvalidEvidence(errors.DefaultCodespace, "empty evidence")
 	}
-	return string(bz)
-}
 
-// Hash implements exported.Evidence interface
-func (ev Evidence) Hash() cmn.HexBytes {
-	return tmhash.Sum(SubModuleCdc.MustMarshalBinaryBare(ev))
+	if err := m.Evidence.ValidateBasic(); err != nil {
+		return err
+	}
+
+	return host.DefaultClientIdentifierValidator(m.ClientID)
 }
 
-// ValidateBasic implements exported.Evidence interface
-func (ev Evidence) ValidateBasic() sdk.Error {
-	if ev.DuplicateVoteEvidence == nil {
-		return sdk.ConvertError(errors.ErrInvalidEvidence(errors.DefaultCodespace, "duplicate evidence is nil"))
-	}
-	err := ev.DuplicateVoteEvidence.ValidateBasic()
-	if err != nil {
-		return sdk.ConvertError(errors.ErrInvalidEvidence(errors.DefaultCodespace, err.Error()))
-	}
-	if ev.ChainID == "" {
-		return sdk.ConvertError(errors.ErrInvalidEvidence(errors.DefaultCodespace, "chainID is empty"))
-	}
-	if ev.ValidatorPower <= 0 {
-		return sdk.ConvertError(errors.ErrInvalidEvidence(errors.DefaultCodespace, fmt.Sprintf("Invalid Validator Power: %d", ev.ValidatorPower)))
-	}
-	if ev.TotalPower < ev.ValidatorPower {
-		return sdk.ConvertError(errors.ErrInvalidEvidence(errors.DefaultCodespace, fmt.Sprintf("Invalid Total Power: %d", ev.TotalPower)))
+// CheckMisbehaviour checks if the evidence provided is a misbehaviour
+func CheckMisbehaviour(trustedCommitter Committer, m Misbehaviour) error {
+	if err := m.ValidateBasic(); err != nil {
+		return err
 	}
-	return nil
-}
 
-// GetConsensusAddress implements exported.Evidence interface
-func (ev Evidence) GetConsensusAddress() sdk.ConsAddress {
-	return sdk.ConsAddress(ev.DuplicateVoteEvidence.Address())
-}
+	trustedValSet := trustedCommitter.ValidatorSet
 
-// GetHeight implements exported.Evidence interface
-func (ev Evidence) GetHeight() int64 {
-	return ev.DuplicateVoteEvidence.Height()
-}
+	// Evidence is on same height as trustedCommiter. ValidatorSets must be the same
+	if trustedCommitter.GetHeight() == uint64(m.GetHeight()) {
+		if !bytes.Equal(trustedValSet.Hash(), m.Evidence.Header1.ValidatorSet.Hash()) ||
```

Comment:
Why is evidence at the same height as `trustedCommiter`?

Reply:
This was an attempt at adding a stricter check on the validatorSets that sign off on misbehaviour that happens to be at the same height (or next height) of a known client update.

My thinking was that if the evidence was far away from the previous update. (e.g. Evidence is on height 30, previous update was on height 20), then Ill have to do the 2/3 power checking thing with the previous validatorset we know.

But if the evidence gets submitted at the same height or the next height as the previous update, i thought we could enforce a tighter bound given that we know both the validatorSet that committed the original update as well as the next validatorSet (since we persist its hash)

However, I don't believe this stricter check is correct anymore since the original update could have been made with an incorrect validator set.

So im going to remove these two conditions